Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/ApromoreZKListener.java

```diff
@@ -0,0 +1,200 @@
+import javax.ws.rs.core.MediaType;
+
+import com.nimbusds.jwt.JWTParser;
+import io.netty.handler.codec.base64.Base64Decoder;
+import org.apromore.plugin.portal.PortalLoggerFactory;
+import org.apromore.portal.security.helper.JwtHelper;
+import org.apromore.portal.util.SecurityUtils;
+import org.slf4j.Logger;
+import org.zkoss.spring.SpringUtil;
+import org.zkoss.zk.ui.Execution;
+import org.zkoss.zk.ui.Session;
+import org.zkoss.zk.ui.event.Event;
+import org.zkoss.zk.ui.event.EventQueues;
+import org.zkoss.zk.ui.util.ExecutionInit;
+
+import java.security.PublicKey;
+import java.util.Base64;
+import java.util.Date;
+
+/**
+ * Callbacks during the lifecycle of ZK scopes.
+ */
+public class ApromoreZKListener implements ExecutionInit {
+
+    private static final Logger LOGGER = PortalLoggerFactory.getLogger(ApromoreZKListener.class);
+    private static final String JWT_COOKIE_NAME = "App_Auth";
+    private static final String KEY_ALIAS = "apseckey";
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation logs the user out if the security token has expired, or refreshes it
+     * if it's still valid but close to expiry.
+     */
+    @Override
+    public void init(final Execution exec, final Execution parent) {
+        LOGGER.debug("Initialize execution {} with parent {}", exec, parent);
+
+        // If there is a parent execution, it will have already performed the required work
+        if (parent != null) {
+            return;
+        }
+
+        final HttpServletRequest httpServletRequest = (HttpServletRequest) exec.getNativeRequest();
+        final HttpServletResponse httpServletResponse = (HttpServletResponse) exec.getNativeResponse();
+
+        // If we are not using Keycloak, we don't have to manage JWTs
+        final ConfigBean config = (ConfigBean) SpringUtil.getBean("portalConfig");
+        final boolean usingKeycloak = config.isUseKeycloakSso();
+        if (!usingKeycloak) {
+            LOGGER.debug("Skipping JWT check because not using Keycloak");
+
+            return;
+        }
+
+        final String appAuthHeader = JwtHelper.readCookieValue(httpServletRequest, JWT_COOKIE_NAME);
+        LOGGER.trace("Read App_Auth cookie: {}", appAuthHeader);
+        final String signedAuthHeader = JwtHelper.readCookieValue(httpServletRequest, JwtHelper.SIGNED_AUTH_HEADER_KEY);
+
+        try {
+            final JWTClaimsSet jwtClaimsSet = JwtHelper.getClaimsSetFromJWT(appAuthHeader);
+
+            final String issuedAtStr = jwtClaimsSet.getStringClaim(JwtHelper.STR_JWT_KEY_ISSUED_AT);
+            LOGGER.debug("issuedAtStr {}", issuedAtStr);
+            final Date issuedAtDate = jwtClaimsSet.getIssueTime();
+            LOGGER.trace("issuedAtDate {}", issuedAtDate);
+            final Date expiryAtDate = jwtClaimsSet.getExpirationTime();
+            LOGGER.debug("expiryAtDate {}", expiryAtDate);
+
+            if (!checkDigitalSignature(appAuthHeader, signedAuthHeader)) {
```

Comment:
Checking the signature is actually a relatively expensive operation, so we'd ultimately prefer not to check it on every ZK Execution as we're currently doing here.  We should be able to optimize this in future by caching the JWT on the user's ZK Session on only doing the redoing signature check if the JWT from the App_Auth cookie has changed.  For now we're proposing to just confirm that everything works otherwise before adding such performance optimizations.

Reply:
The optimization to only check signatures we haven't previously checked has been implemented.